src/routes/communityGroups/community_groups.contoller.js

```diff
@@ -0,0 +1,82 @@
+const fs = require("fs");
+const { promisify } = require("util");
+
+const { ERROR_CODES } = require("../../middleware/error-handler");
+
+const data = require("../../../db/communityGtoups.data.json");
+
+const writeFile = promisify(fs.writeFile);
+
+const dbPath = "db/communityGtoups.data.json";
+
+// return all community groups
+const listCommunityGroups = (req, res) => {
+  return res.status(200).json(data);
+};
+
+// get a community group by its id if it exists
+const getCommunityGroup = (req, res) => {
+  const record = data.filter(
+    record => record.id === parseInt(req.params.id)
+  )[0];
+
+  if (!record) {
+    throw new Error(
+      JSON.stringify({
+        status: ERROR_CODES.NOT_FOUND,
+        message: "Record not found"
+      })
+    );
+  }
+
+  return res.status(200).json(record);
+};
+
+// add a community group if the request has no errors
+const addCommunityGroup = async (req, res) => {
+  const id = data.length + 1;
+  const newCommunityGroupsData = [...data, { id, ...req.body }];
+  await writeFile(dbPath, JSON.stringify(newCommunityGroupsData));
+  return res.status(201).json({
+    id,
+    ...req.body
+  });
+};
+
+// update a communityGroup by its id if the request has no errors
+const updateCommunityGroup = async (req, res) => {
+  let groupData;
+
+  const newCommunityGroupsData = data.map(group => {
+    if (group.id === parseInt(req.params.id)) {
+      groupData = { ...group, ...req.body };
+      return groupData;
+    }
+
+    return group;
+  });
+
+  await writeFile(dbPath, JSON.stringify(newCommunityGroupsData));
+
+  // to avoid overexposure of data, we don't inform if the requested record wasn't available
+  return res.status(200).json(groupData);
+};
+
+// delete a communityGroup by its id if it exists
+const deleteCommunityGroup = async (req, res) => {
+  const newCommunityGroupsData = data.filter(
+    group => group.id !== parseInt(req.params.id)
+  );
+  await writeFile(dbPath, JSON.stringify(newCommunityGroupsData));
+
+  // to avoid overexposure of data, we don't inform if the requested record wasn't available
+  return res.status(200).send();
```

Comment:
There is a more appropriate status code for deleting a record. What is it?

Reply:
204?

I don't know that I really understand when to use vs. the other though?
Documentation I found states 
```
A 204 (No Content) status code if the action has been enacted and no further information is to be supplied.
A 200 (OK) status code if the action has been enacted and the response message includes a representation describing the status.
```

However, 200 doesn't seem to mean "Yes, further action is required", and with the record deleted, I don't know what it's intended by "includes a representation describing the status". What representation?

Reply:
Correct, 204 NO CONTENT is what I had in mind.

You're right that 200 doesn't mean 'further action is required'. The representation is the content of the response body. In RESTful APIs, every request/response body is a representation of some resource record.  In the case of DELETE requests, there is nothing to represent, so the response body should be empty.  If you use the 204 NO CONTENT status, then that tells the client that there is no need to even attempt to parse a body.

That being said, 200 is a valid status response for a DELETE call, but in that case, it's somewhat expected that the body will then contain a representation of the now-deleted record.  This preserves the meaning of 200 to mean 'Success, please parse my body now'.